doc/sphinx/Pacemaker_Explained/constraints.rst

```diff
@@ -434,6 +434,20 @@ Because the above example lets ``symmetrical`` default to TRUE, **Webserver**
 must be stopped before **Database** can be stopped, and **Webserver** should be
 stopped before **IP** if they both need to be stopped.
 
+Symmetric and asymmetric ordering
+_________________________________
+
+A mandatory symmetric ordering of "start A then start B" implies not only that
+the start actions must be ordered, but that B is not allowed to be active
+unless A is active. For example, if the ordering is added to the configuration
+when A is stopped (due to target-role, failure, etc.) and B is already active,
+then B will be stopped.
+
+By contrast, asymmetric ordering of "start A then start B" means the stops can
+occur in either order, which implies that B *can* remain active in the same
+situation.
```

Comment:
It's great to have this documented. However, I still disagree with this behavior, for the reasons discussed in our email thread.

To summarize for posterity: If a new ordering is added while A is stopped and B is active, then we're **only** dealing with a state dependency. There is no action ordering dependency in either direction ("start A then start B" or "stop B then stop A"), since neither resource is starting or stopping when the ordering is added.

Symmetry is irrelevant. The current state (A stopped, B started) could be viewed in either direction. From the perspective that "Rsc A has stopped 'before' rsc B stopped," this is allowed if ordering is asymmetric. But an entirely equally valid perspective is that "Rsc B has started 'before' rsc A started," which is disallowed by our forward ordering constraint.

(I put "before" in quotes because again, it's a matter of perspective. We're dealing only with a state dependency in this example. We're not currently using any knowledge of when resource B started vs. when resource A started... although that could be an option if we check the resource action history, come to think of it.)

Reply:
I agree it's logically questionable, but if this were not the case, asymmetry would lose all effect. If "start A then start B" means B can't be active unless A is active, then B always has to be stopped before A can be stopped.

To put it another way, imagine that A is started, then B is started, then A gets disabled. If B has to stop, what's the point of asymmetric? They both have to be stopped, but the stops can occur in either order? That seems pointless to bother specifying.

Therefore, I think that mandatory symmetric constraints can be considered as state constraints, whereas asymmetric, optional, and serialized constraints are purely action constraints.

I imagine the user's intent for "start A then start B" to be:

* Optional symmetric: "Things go more smoothly in this order if both are starting or stopping, but each works fine without the other"
* Optional asymmetric: "Things go more smoothly in this order if both are starting, but each works fine without the other"
* Mandatory symmetric: "B won't work unless A is active"
* Mandatory asymmetric: "I really really want A to be active first for some reason, but once they're started, each works fine without the other"